src/vmm/src/lib.rs

```diff
@@ -421,23 +422,63 @@ impl Vmm {
             .map(|handle| {
                 handle
                     .response_receiver()
-                    .recv_timeout(Duration::from_millis(400))
+                    .recv_timeout(Duration::from_millis(1000))
             })
             .collect::<std::result::Result<Vec<VcpuResponse>, RecvTimeoutError>>()
-            .map_err(|_| SaveMicrovmStateError::InvalidVcpuState)?;
+            .map_err(|_| UnexpectedVcpuResponse)?;
 
         let vcpu_states = vcpu_responses
             .into_iter()
             .map(|response| match response {
-                VcpuResponse::SaveState(state) => Ok(*state),
-                VcpuResponse::SaveStateFailed(_) => Err(SaveMicrovmStateError::InvalidVcpuState),
-                _ => Err(SaveMicrovmStateError::InvalidVcpuState),
+                VcpuResponse::SavedState(state) => Ok(*state),
+                VcpuResponse::Error(e) => Err(SaveVcpuState(e)),
+                _ => Err(UnexpectedVcpuResponse),
             })
-            .collect::<std::result::Result<Vec<VcpuState>, SaveMicrovmStateError>>()?;
+            .collect::<std::result::Result<Vec<VcpuState>, MicrovmStateError>>()?;
 
         Ok(vcpu_states)
     }
 
+    #[cfg(target_arch = "x86_64")]
+    /// Restores vcpus kvm states.
+    pub fn restore_vcpu_states(
+        &mut self,
+        mut vcpu_states: Vec<VcpuState>,
+    ) -> std::result::Result<(), MicrovmStateError> {
+        use self::MicrovmStateError::*;
+
+        if vcpu_states.len() != self.vcpus_handles.len() {
+            return Err(InvalidInput);
+        }
+        for (handle, state) in self.vcpus_handles.iter().zip(vcpu_states.drain(..)) {
+            handle
+                .send_event(VcpuEvent::RestoreState(Box::new(state)))
+                .map_err(MicrovmStateError::SignalVcpu)?;
+        }
+
+        let vcpu_responses = self
+            .vcpus_handles
+            .iter()
+            // `Iterator::collect` can transform a `Vec<Result>` into a `Result<Vec>`.
+            .map(|handle| {
+                handle
+                    .response_receiver()
+                    .recv_timeout(Duration::from_millis(1000))
```

Comment:
I've increased all timeouts to 1000ms to give a better chance to highly oversubscribed scenarios/machines.

We've seen scenarios in the past where people were running stress tests and where channel timeouts would happen simply because CPUs were so contended that threads of a particular Firecracker process were scheduled to run very sparsely.

We do not enforce in Firecracker a time-limit for these operations, they will run only as fast as the host CPU runs them. The timeout's purpose here is to catch error cases where things go wrong and the thread on the other end of the channel is not responding anymore.

I'm not sure even 1000ms is enough, but it is what was used to fix https://github.com/firecracker-microvm/firecracker/issues/1555 so we can continue with that for now.